File: CPP/14_longestCommonPrefix/14_longestCommonPrefix.cpp
// 14_longestCommonPrefix.cpp : 定义控制台应用程序的入口点。
//

#include "stdafx.h"
#include <iostream>
#include <string>
#include <vector>
using namespace std;

string longestCommonPrefix(std::vector<string> strs)
{
	string str;
	if (strs.size() == 0)
	{
	}
	else if (strs.size() == 1)
	{
		str = strs[0];
	}
	else
	{
		unsigned int i = 0;
		for (; i < strs[0].length(); i++)
		{
			str = strs[0].substr(0, i + 1);

			unsigned int j = 1;
			for (; j < strs.size(); j++)
			{
				if (str != strs[j].substr(0, i + 1))
				{
					break;
				}
			}

			if (j != strs.size())
			{
				str = str.substr(0, i);
				break;
			}
		}
	}

	return str;
}

int _tmain(int argc, _TCHAR* argv[])
{
	string strs1[] = {
		"flower",
		"flow",
		"flight"
	};

	cout << longestCommonPrefix(std::vector<string>(strs1, strs1 + sizeof(strs1) / sizeof(strs1[0]))) << endl;
	system("pause");
	return 0;
}

File: Go/14_longestCommonPrefix/14_longestCommonPrefix.go
package q014

import (
	"fmt"
	"strings"
)

func longestCommonPrefix(strs []string) string {
	l := len(strs)
	if l == 0 {
		return ""
	} else if l == 1 {
		return strs[0]
	}

	// horizontal
	lcp := strs[0]
	for i := 1; i < l; i++ {
		for strings.Index(strs[i], lcp) != 0 {
			lcp = lcp[0: len(lcp) - 1]
			if lcp == "" {
				return ""
			}
		}
	}

	// vertical
	for i := 0; i < len(strs[0]); i++ {
		ch := strs[0][i]
		for j := 1; j < l; j++ {
			if len(strs[j]) < i + 1 || strs[j][i] != ch {
				return strs[0][0: i]
			}
		}
	}

	return lcp
}

// Test ...
func Test() {
	fmt.Println("14_longestCommonPrefix begin")
	strs := []string{
		"flower",
		"flow",
		"flight",
	}
	fmt.Printf("%v\n", longestCommonPrefix(strs))
	fmt.Println("14_longestCommonPrefix end")
}
